### PR TITLE
Implement periodic writing of alertmanager state to storage.

### PR DESCRIPTION
**Pull request by stevesg**
**What this PR does**:
When ring-based/sharding replication is enabled, the alertmanager state
(silences, notification log) is periodically written to object storage
so that it can be used to recover from an all-replica outage. Only one
of the replicas is responsible for writing the state (position 0).

**Checklist**
- [X] Tests updated
- [X] ~~Documentation added~~
- [X] ~~`CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`~~

### CODE REVIEW

File: pkg/alertmanager/state_persister.go
Comment (pracucci) [inline pkg/alertmanager/state_persister.go:17]: I would make this configurable (via CLI flag / YAML config). It's something a Cortex cluster operator may need to tune.

File: pkg/alertmanager/state_persister.go
Comment (pracucci) [inline pkg/alertmanager/state_persister.go:40]: I think we need some metrics to track the health of all state persisters. Eg. state persistence attempts + failures counter. Without metrics we can't alert on failures in the persister.

File: pkg/alertmanager/state_persister.go
Comment (pracucci) [inline pkg/alertmanager/state_persister.go:62]: You can simplify this using `services.NewTimerService()` instead of `services.NewBasicService()`.

**Review by pracucci — COMMENTED** [review]
The logic makes sense to me! I left a couple of nits. Waiting for the final PR to do a deeper review but I haven't seen any issue so far 👏

File: pkg/alertmanager/state_persister.go
Comment (stevesg) [reply pkg/alertmanager/state_persister.go]: Agreed. I will add these in a separate PR to keep the size of this PR small, and I would also like to add some other metrics relating to the state (e.g. number of times we had read the state from a replica. number of times reading from storage, and so forth).

File: pkg/alertmanager/multitenant.go
Comment (pracucci) [inline pkg/alertmanager/multitenant.go:164]: Please explicitly mention this is used only when alertmanager sharding is enabled.

Comment (pracucci) [inline pkg/alertmanager/multitenant.go:164]: [nit] Given you've created the `PersisterConfig` you may do a step forward and move the `RegisterFlags()` and its `Validate()` directly on `PersisterConfig`. Then you can include it in the `MultitenantAlertmanagerConfig` without creating a subfield with:

```
	// For the state persister.
	PersistInterval time.Duration `yaml:"-"`
```

**Review by pracucci — APPROVED** [review]
Good job! LGTM (modulo a couple of nits)

File: docs/configuration/config-file-reference.md
Comment (gotjosh) [inline docs/configuration/config-file-reference.md:2085]: > his state is read when all replicas for a shard have failed. 

This might be a little confusing - We'll read state from storage in the case of this replica being unable to communicate with the other e.g. transient network failures. Under that scenario, the other replicas are completely fine. 

Is it worth clarifying that it is not on actual replica failure but on the lack of a proper response from them?

**Review by gotjosh — APPROVED** [review]

File: docs/configuration/config-file-reference.md
Comment (ranton256) [reply docs/configuration/config-file-reference.md:2085]: +1, I think it probably makes sense to say "when all replicas for a shard can not be contacted" instead, right, since won't know the cause, failure or network problem.

**Review by ranton256 — APPROVED** [review]
Everything looks good to me, just besides the minor point of the wording discussion about the documentation part.

Thanks!

**Review by pstibrany — APPROVED** [review]
LGTM. Nice small readable PR!

File: docs/configuration/config-file-reference.md
Comment (stevesg) [reply docs/configuration/config-file-reference.md:2085]: Agreed, I'll update this.

**Review by pracucci — APPROVED** [review]
Thanks for addressing my feedback! 🙏

**Review by gotjosh — APPROVED** [review]

**Pull request merged**
